pkg/agent/nodeportlocal/portcache/port_table_linux.go

```diff
@@ -44,38 +44,34 @@ var (
 )
 
 func (pt *PortTable) GetEntry(ip string, port int, protocol string) *NodePortData {
-	var _ = protocol
 	pt.tableLock.RLock()
 	defer pt.tableLock.RUnlock()
 	// Return pointer to copy of data from the PodEndpointTable.
-	if data := pt.getEntryByPodIPPort(ip, port); data != nil {
+	if data := pt.getEntryByPodIPPortProto(ip, port, protocol); data != nil {
 		dataCopy := *data
 		return &dataCopy
 	}
 	return nil
 }
 
-func openSocketsForPort(localPortOpener LocalPortOpener, port int) ([]ProtocolSocketData, error) {
-	// Port needs to be available for all supported protocols: we want to use the same port
-	// number for all protocols and we don't know at this point which protocols are needed.
-	// This is to preserve the legacy behavior of allocating the same nodePort for all protocols.
-	protocols := make([]ProtocolSocketData, 0, len(supportedProtocols))
-	for _, protocol := range supportedProtocols {
-		socket, err := localPortOpener.OpenLocalPort(port, protocol)
-		if err != nil {
-			klog.V(4).InfoS("Local port cannot be opened", "port", port, "protocol", protocol)
-			return protocols, err
-		}
-		protocols = append(protocols, ProtocolSocketData{
-			Protocol: protocol,
-			State:    stateOpen,
-			socket:   socket,
-		})
+func openSocketsForPort(localPortOpener LocalPortOpener, port int, protocol string) ([]ProtocolSocketData, error) {
```

Comment:
Shall we return type `ProtocolSocketData` instead of a slice since there is one socket opened at most?

Reply:
Even for type `NodePortData`, do we still need a slice of `ProtocolSocketData`? Maybe one object is OK?

Reply:
Updated.